README
# file:   README - Notes on TitanCNA version updates
# author: Gavin Ha 
# 		  Dana-Farber Cancer Institute
#		 	  Broad Institute
# contact: <gavinha@gmail.com> or <gavinha@broadinstitute.org>
# TitanCNA website: http://compbio.bccrc.ca/software/titan/
# Google Groups: https://groups.google.com/forum/#!forum/titancna 
# date:	  September 26, 2016
# description: TitanCNA is a probabilistic framework for analyzing subclonal copy number alterations (CNA) and loss of heterozygosity (LOH) in whole genome sequencing of tumours. The main software is implemented as a R Bioconductor package with preprocessing tools implemented in Python and BAMtools (C++). The example provided will reproduce Figure 1 in the manuscript. However, it will be slightly different because the example is only based the analysis of chr2, not genome-wide.

The license for the R Bioconductor package is now open source under GPLv3.  This applies to the v1.9.0 and all subsequent versions within and obtained from Bioconductor.  
Because the Bioconductor SVN is bridged to the GitHub TitanCNA repository "master" branch, the GitHub master branch for TitanCNA will also be under GPLv3.  
Please note that any other branch in the GitHub TitanCNA repository is still under the previous academic license. Users who are using TitanCNA for the purpose of academic research are free to use all branches and versions of the software. If this does not apply to you, please contact gavinha@broadinstitute.org, sshah@bccrc.ca, and prebstein@bccancer.bc.ca.

Thank you for your interest in the TitanCNA software.

=================================================================================
TitanCNA version 1.10.1 changes 
Changed to v1.12.1 to match Bioconductor versioning
=================================================================================
1) New function
	- outputTitanSegments(): Replaces the need for external Perl script to generate segments file.
	
2) Modified function
	- setupClonalParameters(): Initializes cellular prevalence values evenly distributed between 0 and 1 rather than 0.1 increments.
	
	- loadDefaultParameters(): added hetBaselineSkew argument which will determine the baseline allelic fraction to use.  If this is NULL, then will set using the median allelic ratio across the genome ("data" argument).  If "data" argument is NULL, then will simply set to 0.55 (heterozgyous skew of 0.05). Note that the "skew" argument applies to all other states except for the heterozygous allelic states (e.g. 1:1, 2:2, 3:3, 4:4 - for minor:major copies).

3) Bug fixes
	- plotClonalFrequency(): Handle case when result is completely copy neutral heterozygous and cellular prevalence and ClonalClusters columns are both NA
	
	- runEMclonalCN(): If EM ever decreases, then take last iteration (with higher likelihood).
	
	- filterData(): Remove centromere positions now working. 

=================================================================================
TitanCNA version 1.7.2 changes (Committed revisions 6796790)
Merged to master as v1.9.0 to match Bioconductor versioning
=================================================================================
1) New function to remove clonal clusters with few or no data
	- removeEmptyClusters(): This function can help prune down the number of clonal clusters for which there are few alterations. 
2) Updated functions
	- plotCNlogRByChr(): added segment plotting when given list of segments 
	- plotCNlogRByChr(): accepts normal contamination estimate so that the average tumor ploidy adjustment is correctly tumor-specific.  
3) Updated dependencies
	- R (>= 3.2.0), foreach (>= 1.4.2), IRanges (>= 2.2.4), Rsamtools (>= 1.20.4), GenomeInfoDb (>= 1.4.0)
4) New dependencies
	- GenomicRanges (>= 1.20.4)
5) Bioconductor package license changed to GPLv3.

=================================================================================
TitanCNA version 1.5.8 changes (Committed revisions 977c48a,5fe5acc.)
Merged to master as v1.7.1 (a5bf4f3)
=================================================================================
1) Updated functions to align with dependent package updates.
	- Affected function: loadAlleleCounts, correctReadDepth
	- Adjusted function to accommodate GenomeInfoDb v1.1.3 -> v1.2.4
	- Usage of this function remains unchanged.

2) Minor bug fix in function
	- Affected function: correctReadDepth
	- Issue with findOverlaps() usage fixed

3) Bug fix
	- Affected function: outputTitanResults, plotClonalFrequency, plotSubcloneProfiles
	- Fixed bug when error is thrown trying to output results for genome that is predicted to be completely diploid HET.

4) Modified internal functionality
	- Affected function: runEMclonalCN, viterbiClonalCN
	- The C code (src/fwd_backC_clonalCN.c & src/viterbiC_clonalCN.c) have been modified such that the transition probabilities between clonal clusters for HET (diploid) states are equal.  This may help prevent situations when normal contamination is 0 for more than 1 clonal cluster solutions.

=================================================================================
TitanCNA version 1.5.7 changes (Committed revision 98539.)
=================================================================================
1) Fixed bug in utils.R, src/fwd_backC_clonalCN.c, and src/viterbiC_clonalCN.c
	- unfinished changes were accidentally committed in v1.5.6.
	- rolled "loadDefaultParameters" back to v1.5.5

=================================================================================
TitanCNA version 1.5.6 changes (Committed revision 98115.)
=================================================================================
1) Fixed bug in getPositionOverlap
	- Affected function: getPositionOverlap
	- This function should return log ratio values based on overlap of SNPs ("chr" & "posn") with the input list of regions "dataVal". The bug causes the return of values that are incorrectly matched to the input coordinates. This bug is more prominent when using targeted sequencing (i.e. exome) data. In some cases, this can lead to missing data and/or chromosomes in the final result. This has been fixed.

2) Added option to remove SNPs near centromeres
	- Affected function: filterData
	- This function can now remove SNPs at and/or near centromeres. The "centromeres" argument (default NULL) accepts a data.frame with 3 columns: chromosome, start, stop.  Argument "centromere.flankLength" (default 0) accepts an integer specifying the length of region to the left and right of the centromere to also exclude overlapping SNPs.
	
3) Updated vignette to include latest changes from v1.5.5. The modification of loadDefaultParameters to include the loaded data from loadAllelicCounts to inform the baseline allelic ratio parameters.

4) Modified documentation manuals
	- Affected man page: loadDefaultParameters
	- Fixed language and grammar, and improved clarity

=================================================================================
TitanCNA version 1.5.5 changes (Committed revision 96727.)
=================================================================================
1) Modified function
	- Affected function: loadDefaultParameters
	- Added argument: data
	- For symmetric=TRUE, the heterozygous states (4,9,25) are set to the median allelic ratio of the entire genome if data is given
	- data is the output from loadAlleleCounts()
	- This corrects the issue where the entire genome is mostly copy neutral LOH (NLOH) when it should most likely be copy neutral heterozygous (HET).
	
2) Modified function
	- Affected function: outputModelParameters
	- Argument removed: S_Dbw.data.type
	- Will output S_Dbw for datatypes: LogRatio, AllelicRatio, and Both
	- "Both" refers to the sum of both datatypes (LogRatio and AllelicRatio) for each of the dens.bw and scat components 

=================================================================================
TitanCNA version 1.5.4 changes (Committed revision 96687.)
=================================================================================
1) Model selection default changed
	- Affected function: outputModelParameters
	- New value can be used for argument: S_Dbw.data.type = "Both"
	- "Both" will sum the S_Dbw index for "AllelicRatio" and "LogRatio"; this will account for both datatypes when selecting the best solution
	- outputModelParameters() will now return the S_Dbw information
	
2) Vignette additions
	- Under section "Format and print results to file"
		- added instructions for using S_Dbw scale to penalize higher number of clonal clusters
		
3) Updated references in manual pages
	- Ha et al. (2014). Genome Research, 24: 1881-1893.
	
=================================================================================
TitanCNA version 1.5.3 changes (Committed revision 96468.)
=================================================================================
Note: This is the development version
1) Minor implementation change
	- Affected function: viterbiClonalCN, runEMclonalCN
	- Probabilities for transitioning into different clonal clusters was originally txnZstrength*txnExpLen and specified in the C code.  It is now decoupled from the C code and specified in the R code, instead. 

2) Vignette correction/additions
	- Under section "Filter the data", changed to:
		- data <- filterData(data, c(1:22, "X", "Y"),...
	- Under section "Format and print results to file"
		- added instructions for using S_Dbw scale to penalize higher number of clonal clusters

3) Bug fixes
	- Affected function: plotSubcloneProfiles
	- Whole genome profile plots of subclones (using chr=NULL) was not working properly, previously.
	
=================================================================================
TitanCNA version 1.5.2 changes (Committed revision 95883.)
=================================================================================
Note: Bioconductor v3.1 (development)
1) Modified functions
	- Affected function: getPositionOverlap
	- Now uses RangedData objects and findOverlaps utility function internally

NEWS
# file:   README - Notes on TitanCNA version updates
# author: Gavin Ha 
# 		  Dana-Farber Cancer Institute
#		  Broad Institute
# contact: <gavinha@gmail.com> or <gavinha@broadinstitute.org>
#
# TitanCNA GitHub: https://github.com/gavinha/TitanCNA
# Snakemake Workflow: https://github.com/gavinha/TitanCNA/tree/master/scripts/snakemake
# Google Groups: https://groups.google.com/forum/#!forum/titancna 
# TitanCNA website: http://compbio.bccrc.ca/software/titan/
# TitanCNA KRONOS workflow: https://github.com/MO-BCCRC/titan_workflow

# date:	  March 22, 2018
#
# description: TitanCNA is a probabilistic framework for analyzing subclonal copy number alterations (CNA) and loss of heterozygosity (LOH) in whole genome sequencing of tumours. The main software is implemented as a R Bioconductor package with preprocessing tools implemented in Python and BAMtools (C++). The example provided will reproduce Figure 1 in the manuscript. However, it will be slightly different because the example is only based the analysis of chr2, not genome-wide.

Thank you for your interest in the TitanCNA software.

=================================================================================
TitanCNA version 1.17.1 changes 
=================================================================================
1)  New functions: 
	- correctIntegerCN(): recomputes high-level copy number that is capped by the maximum CN state. 
	performs two tasks - (1) correct log ratio based on purity and ploidy, and then convert to decimal CN value; (2) Correct bins and segments in which the original predicted integer copy number was assigned the maximum CN state; bins and segments for all of chromosome X are also corrected, if provided in the input.

2) Modified functions:
	- plotSegmentMedians() and plotCNlogRByChr(): includes argument to show color-coding for corrected copy number; defaults to TRUE for this argument.
	
3) Removed functions/manual/dependencies:
	- extractAlleleReadCounts()	
	- Rsamtools dependency

=================================================================================
TitanCNA version 1.15.0 changes 
=================================================================================
1) 10X Genomics analysis
  - Please see https://github.com/gavinha/TitanCNA/tree/10X_Genomics/scripts for instructions on running the 10X Genomics data

2) New functionality
  - plotSegmentMedians()
  - loadHaplotypeAlleleCounts(): loads input allele counts with phasing information
  - plotHaplotypeFraction(): results from 10X Genomics WGS data with phasing of haplotype blocks
  - loadDefaultParameters(): now accepts the argument "alleleEmissionModel" to specify "binomial" for standard WGS depth data or "Gaussian" for haplotype counts (summarized over a given bin size - see loadHaplotypeAlleleCounts())
  
3) Modified functions
  -filterData: will remove chromosomes with a single data point to avoid the error (task 23 failed - "INTEGER() can only be applied to a 'integer', not a 'NULL'")

4) Modified features (no changes for user-accessible functions)
  - updateParameters: coordinate descent estimate of ploidy update uses previously estimated normal parameter from the same corodinate descent iteration and leads to faster convergence


=================================================================================
TitanCNA version 1.10.1 changes 
Changed to v1.12.1 to match Bioconductor versioning
=================================================================================
1) New function
	- outputTitanSegments(): Replaces the need for external Perl script to generate segments file.
	
2) Modified function
	- setupClonalParameters(): Initializes cellular prevalence values evenly distributed between 0 and 1 rather than 0.1 increments.
	
	- loadDefaultParameters(): added hetBaselineSkew argument which will determine the baseline allelic fraction to use.  If this is NULL, then will set using the median allelic ratio across the genome ("data" argument).  If "data" argument is NULL, then will simply set to 0.55 (heterozgyous skew of 0.05). Note that the "skew" argument applies to all other states except for the heterozygous allelic states (e.g. 1:1, 2:2, 3:3, 4:4 - for minor:major copies).

3) Bug fixes
	- plotClonalFrequency(): Handle case when result is completely copy neutral heterozygous and cellular prevalence and ClonalClusters columns are both NA
	
	- runEMclonalCN(): If EM ever decreases, then take last iteration (with higher likelihood).
	
	- filterData(): Remove centromere positions now working. 

=================================================================================
TitanCNA version 1.7.2 changes (Committed revisions 6796790)
Merged to master as v1.9.0 to match Bioconductor versioning
=================================================================================
1) New function to remove clonal clusters with few or no data

R/haplotype.R
#' author: Gavin Ha 
#' 		Dana-Farber Cancer Institute
#'		Broad Institute
#' contact: <gavinha@gmail.com> or <gavinha@broadinstitute.org>
#' date:	  January 11, 2018

#' @import data.table
#' @import GenomicRanges

loadBXcountsFromBEDDir <- function(bxDir, chrs = c(1:22, "X", "Y"), minReads = 2){
	files <- list.files(bxDir, pattern=".bed", full.names = TRUE)
	message("Loading BX counts from ", bxDir)
	bxAll <- NULL
	for (i in files){
		message(i)
		awkcmd <- paste0("awk -F \"\t\" \'{if (!$4) {print $1\"\t\"$2\"\t\"$3\"\t\"\"NA\"} else {print $1\"\t\"$2\"\t\"$3\"\t\"$4}}\' ", i)
		bxChr <- fread(awkcmd, header = FALSE, na.strings = c("NA"))
		colnames(bxChr) <- c("chr", "start", "end", "BX")
		bxChr[, BXcounts:=sapply(BX, function(x){
		if (!is.na(x)){
				bxCodes <- unlist(tstrsplit(x,","))
				bxReads <- as.numeric(tstrsplit(bxCodes, "_")[[2]])
				return(sum(bxReads >= minReads))
			}else{
				return(0)
			}
		})]
		bxChr[, BX:=NULL]
		bxAll <- rbind(bxAll, bxChr)
	}
	bxGR <- RangedData(space = bxAll[[1]], ranges = IRanges(start = bxAll[[2]], 
										 end = bxAll[[3]]), BX.count = bxAll[[4]])
	bxGR <- keepChr(bxGR, chr = chrs)
	return(bxGR)
}


loadReadCountsFromBed <- function(counts, chrs = c(1:22, "X", "Y"), gc = NULL, map = NULL, centromere = NULL, flankLength = 100000, targetedSequences = NULL, genomeStyle = "NCBI"){

	names(counts) <- setGenomeStyle(names(counts), genomeStyle)
	counts <- keepChr(counts, chrs)
	if (!is.null(gc)){ 
		names(gc) <- setGenomeStyle(names(gc), genomeStyle)
		gc <- keepChr(gc, chrs)
		counts$gc <- gc$value
	}
	if (!is.null(map)){ 
		names(map) <- setGenomeStyle(names(map), genomeStyle)
		map <- keepChr(map, chrs)
		counts$map <- map$value
	}
	colnames(counts)[1] <- c("reads")
	
	# remove centromeres
	if (!is.null(centromere)){ 
		counts <- excludeCentromere(counts, centromere, flankLength = flankLength)
	}
	# keep targeted sequences
	if (!is.null(targetedSequences)){
		countsExons <- filterByTargetedSequences(counts, targetedSequences)
		counts <- counts[countsExons$ix,]
	}
	## filter 0 read bins ##
	#counts <- counts[counts$reads > 0, ]
	return(counts)
}

loadHaplotypeAlleleCounts <- function(inCounts, fun = "sum", haplotypeBinSize = 1e5, 
      minSNPsInBin = 3, chrs = c(1:22, "X"), minNormQual = 200, 
      genomeStyle = "NCBI", sep = "\t", header = TRUE, seqinfo = NULL) {
	if (is.character(inCounts)){
    ## LOAD INPUT READ COUNT DATA 
    	message("titan: Loading data and phasing information ", inCounts)
    	data <- read.delim(inCounts, header = header, stringsAsFactors = FALSE, 
        		sep = sep)
      colnames(data) <- c("chr", "posn", "refBase", "ref", "nonRefBase", "nonRef", "normQual", "genotype", "phaseSet")
      if (typeof(data[,"posn"])!="integer" || typeof(data[,"ref"])!="integer" || 
          typeof(data[,"nonRef"])!="integer" || is.null(data$genotype) || is.null(data$phaseSet)){
        stop("loadHaplotypeAlleleCounts: Input counts file format does not 
          match required specifications.")		
      }
  }else if (is.data.frame(inCounts)){  #inCounts is a data.frame
    data <- inCounts
  }else{
    stop("loadHaplotypeAlleleCounts: Must provide a filename or data.frame to inCounts")
  }
  
  if (is.null(seqinfo)){
    seqinfo <- readRDS(system.file("extdata", "Seqinfo_hg19.rds", package = "TitanCNA"))
  }
  # convert to desired genomeStyle and only include autosomes, sex chromosomes
  data[, 1] <- setGenomeStyle(data[, 1], genomeStyle)
   
  ## sort chromosomes
	indChr <- orderSeqlevels(as.character(data[, "chr"]), X.is.sexchrom = TRUE)
	data <- data[indChr, ]
	## sort positions within each chr
	for (x in unique(data[, "chr"])){
		ind <- which(data[, "chr"] == x)
		data[ind, ] <- data[ind[order(data[ind, "posn"], decreasing = FALSE)], ]
	}
  
  ## filter data ##
  data <- cbind(data, start = data$posn, end = data$posn, depth = data$ref + data$nonRef)
  data <- data[data$normQual >= minNormQual, ]
  # get max of allele counts
  #data$refOriginal <- data$ref
  #data$ref <- pmax(data$refOriginal, data$nonRef)
  ## use GRanges to find overlap of tiling haplotypeBinSize ##
  phasedAlleles <- getPhasedAlleleFraction(data)
  data$phasedAlleleFraction <- phasedAlleles$allele.fraction
  data$phasedCount <- phasedAlleles$phasedCount
  data$ref.symmetric <- pmax(data$ref, data$nonRef)
  data.gr <- makeGRangesFromDataFrame(data, keep.extra.columns = TRUE, seqinfo = seqinfo, ignore.strand = TRUE)
  tile.gr <- unlist(tileGenome(seqinfo, tilewidth = haplotypeBinSize))
  data.gr <- keepSeqlevels(data.gr, chrs)
  tile.gr <- keepSeqlevels(tile.gr, chrs)
  hits <- findOverlaps(query = data.gr, subject = tile.gr)
  data.gr$haplotypeBin <- subjectHits(hits)
  ## use data.table to process haplotype counts by blocks
  data <- as(data.gr, "data.frame")
  data.dt <- data.table(as(data.gr, "data.frame"))
  haploBinSummary <- data.dt[, list(HaplotypeFraction = mean(phasedAlleleFraction), 
      HaplotypeDepth.sum = sum(phasedCount), HaplotypeBinDepth.sum = sum(depth),
      HaplotypeDepth.mean = round(mean(phasedCount)), #round?
      #HaplotypeDepth.mean = round(mean(ref.symmetric)), 
      HaplotypeBinDepth.mean = round(mean(depth)), #round?
      SNPs = length(phasedAlleleFraction)), by = c("phaseSet", "haplotypeBin")]
  # filter bins by number of SNPs #
  haploBinSummary <- haploBinSummary[SNPs >= minSNPsInBin]
  # summary bins with multiple phaseset ID such that haplotypeBin is unique
  haploBinSummary.unique <- haploBinSummary[, list(SNPs = sum(SNPs),
    HaplotypeFraction = sum(HaplotypeFraction * SNPs) / sum(SNPs), 
    HaplotypeDepth.sum = sum(HaplotypeDepth.sum),
    #HaplotypeDepth.sum.symmetric = sum(HaplotypeDepth.sum.symmetric), 
    HaplotypeBinDepth.sum = sum(HaplotypeBinDepth.sum),
    HaplotypeDepth.mean = round(sum(HaplotypeDepth.mean * SNPs) / sum(SNPs)), #round?
    #HaplotypeDepth.mean.symmetric = round(sum(HaplotypeDepth.mean.symmetric * SNPs) / sum(SNPs)), 
    HaplotypeBinDepth.mean = round(sum(HaplotypeBinDepth.mean * SNPs) / sum(SNPs)), #round?
    phaseSet = phaseSet[which.max(SNPs)]), by = haplotypeBin]
    # get symmetric haplotype fraction
  haploBinSummary.unique[, HaplotypeFraction.symmetric := pmax(HaplotypeFraction, 1 - HaplotypeFraction)]
  haploBinSummary.unique[, HaplotypeDepth.sum.symmetric := pmax(HaplotypeDepth.sum, HaplotypeBinDepth.sum - HaplotypeDepth.sum)]
  haploBinSummary.unique[, HaplotypeDepth.mean.symmetric := pmax(HaplotypeDepth.mean, HaplotypeBinDepth.mean - HaplotypeDepth.mean)]
   # set bin column as key so that we can map back to original data
  setkey(haploBinSummary.unique, haplotypeBin) 
  # add the bin summarized values back to data.dt
  data.dt <- cbind(data.dt, select(haploBinSummary.unique[.(subjectHits(hits))], 
      haplotypeBin.aggr = haplotypeBin, HaplotypeFraction.symmetric, HaplotypeDepth.sum,
      HaplotypeDepth.sum.symmetric, HaplotypeBinDepth.sum, HaplotypeDepth.mean, 
      HaplotypeDepth.mean.symmetric, HaplotypeBinDepth.mean, 
      SNPs, phaseSet.aggr = phaseSet))
  data.dt <- na.omit(data.dt)
  data.dt[, phasedCount.haploSymmetric := {
    if (HaplotypeDepth.sum != HaplotypeDepth.sum.symmetric){
      depth - phasedCount 
    }else{
      phasedCount
    }
  }, by=1:nrow(data.dt)]
  alleleData <- select(data.dt, chr=seqnames, posn=start, 
      refOriginal=ref, nonRef=nonRef, tumDepth=depth)
  alleleData$chr <- as.character(alleleData$chr)
  alleleData$ref = pmax(alleleData$refOriginal, alleleData$nonRef)
  haplotypeData <- select(data.dt, chr=seqnames, posn=start, 
      phaseSet=phaseSet.aggr, refOriginal=ref, tumDepthOriginal = depth)
  haplotypeData$chr <- as.character(haplotypeData$chr)
  if (fun == "sum"){
    haplotypeData$ref <- data.dt[, HaplotypeDepth.sum.symmetric]
    haplotypeData$tumDepth <- data.dt[, HaplotypeBinDepth.sum]
    haplotypeData[, HaplotypeRatio := ref / tumDepth]
    #haplotypeData$haplotypeCount <- data.dt[, phasedCount]#data.dt[, HaplotypeDepth.sum]
    haplotypeData$haplotypeCount <- data.dt[, phasedCount.haploSymmetric]
  }else if (fun == "mean"){
    haplotypeData$ref <- data.dt[, HaplotypeDepth.mean.symmetric]
    haplotypeData$tumDepth <- data.dt[, HaplotypeBinDepth.mean]
    haplotypeData$HaplotypeRatio <- data.dt[, HaplotypeDepth.sum.symmetric] / data.dt[, HaplotypeBinDepth.sum]
    #haplotypeData[, HaplotypeRatio := ref / tumDepth]
    #haplotypeData$haplotypeCount <- data.dt[, phasedCount]#data.dt[, HaplotypeDepth.mean]
    haplotypeData$haplotypeCount <- data.dt[, phasedCount.haploSymmetric]
  }else if (fun == "SNP"){
    haplotypeData$ref <- data.dt[, phasedCount.haploSymmetric]
    haplotypeData$tumDepth <- data.dt[, depth]
    haplotypeData$HaplotypeRatio <- data.dt[, HaplotypeDepth.sum.symmetric] / data.dt[, HaplotypeBinDepth.sum]
    haplotypeData$haplotypeCount <- data.dt[, phasedCount.haploSymmetric]
  }
  haplotypeData$nonRef <- haplotypeData$tumDepth - haplotypeData$ref
  return(list(haplotypeData=haplotypeData, alleleData=alleleData, data=data.dt))
}

getPhasedAlleleFraction <- function(x){
  altInd <- as.logical(as.numeric(tstrsplit(x$genotype, "\\||\\/")[[1]]))
  refInd <- !altInd
  allele.fraction <- rep(NA, length(x))
  allele.fraction[refInd] <- x[refInd, "ref"] / x[refInd, "depth"]
  allele.fraction[altInd] <- x[altInd, "nonRef"] / x[altInd, "depth"]
  phasedCount <- rep(NA, length(x))
  phasedCount[refInd] <- x[refInd, "ref"]
  phasedCount[altInd] <- x[altInd, "nonRef"]
  return(list(allele.fraction = allele.fraction, phasedCount = phasedCount))
}

getHaplotypesFromVCF <- function(vcfFile, chrs = c(1:22, "X"), build = "hg19",
                                 filterFlags = c("PASS", "10X_RESCUED_MOLECULE_HIGH_DIVERSITY"), 
                                 minQUAL = 100, minDepth = 10, minVAF = 0.25, altCountField = "AD",
                                 keepGenotypes = c("1|0", "0|1", "0/1"), snpDB = NULL){
  #require(data.table)
  message("Loading ", vcfFile)
  vcf <- readVcf(vcfFile, genome = build)
  chrName <- mapSeqlevels(seqlevels(vcf), style="NCBI")
  rowRanges(vcf) <- renameSeqlevels(rowRanges(vcf), na.omit(chrName))
  #keepGenotypes = c("1|0", "0|1", "0/1")
  
  ## filter vcf ##
  message("Filtering VCF ...")
  # keep by filter flags
  indFILTER <- rowRanges(vcf)$FILTER %in% filterFlags
  # keep SNPs - ref and alt have length of 1 and only a single allele for ref/alt
  indSNP <- nchar(unstrsplit(CharacterList(rowRanges(vcf)$ALT), sep=",")) == 1 & 
    nchar(unstrsplit(rowRanges(vcf)$REF, sep=",")) == 1
  message("  by quality >=", minQUAL, ")")
  indQUAL <- rowRanges(vcf)$QUAL >= minQUAL
  # keep genotypes
  message("  by genotypes: ", paste0(keepGenotypes, collapse=";"))
  indGeno <- geno(vcf)$GT %in% keepGenotypes
  ind <- indFILTER & indSNP & indQUAL & indGeno
  vcf <- vcf[which(ind)]
  
  message("  by depth (>=", minDepth, ")")
  depth <- geno(vcf)$DP
  indDP <- depth >= minDepth
  message("  by VAF (>=", minVAF, ")")
  minVAF <- min(minVAF, 1 - minVAF) # symmetric from 0-0.5
  if (!is.null(geno(vcf)[[altCountField]])){
    altCounts <- unlist(lapply(geno(vcf)[[altCountField]], min))
  }else{
    stop("Alternate read counts not in AD or AO fields.")
  }
  indVAR <- (altCounts / depth) >= minVAF
  vcf <- vcf[which(indDP & indVAR)]
  rm(depth)
  
  if (!is.null(snpDB)){
    message (" by SNP VCF file ", snpDB)
    snp <- readVcf(snpDB, genome = build)
    snpInd <- nchar(unstrsplit(CharacterList(rowRanges(snp)$ALT), sep=",")) == 1 & 
      nchar(unstrsplit(rowRanges(snp)$REF, sep=",")) == 1
    snp <- snp[which(snpInd)]
    hits <- findOverlaps(query = rowRanges(vcf), subject = rowRanges(snp))
    indSNPDB <- queryHits(hits)
    vcf <- vcf[indSNPDB]
  }
  ## 
  #acounts <- do.call(rbind, geno(vcf)$AD)
  # phased snps
  #indPhased <- grepl(pattern = "\\|", geno(vcf)$GT)
  # phase sets
  #ps.rle <- rle(as.integer(geno(vcf)$PS))
  
  geno.gr <- rowRanges(vcf)	
  values(geno.gr) <- cbind(values(geno.gr), 
                           DataFrame(GT = as.character(geno(vcf)$GT), PS = as.character(geno(vcf)$PS)))
  HT <- getPhasedAllele(geno.gr)
  values(geno.gr) <- cbind(values(geno.gr), DataFrame(HT1 = HT$h1, HT2 = HT$h2))
  geno.gr <- keepSeqlevels(geno.gr, chrs)
  
  return(list(vcf.filtered = vcf, geno = geno.gr))
}

# x GRanges object with GT, REF, ALT metadata columns
# returns reference or alternate allele in haplotype
getPhasedAllele <- function(x){
	h1 <- rep(NA, length(x))
	h2 <- rep(NA, length(x))
	# haplotype 1 (h1) ref allele for 1|0
	h1[x$GT == "1|0"] <- as.character(x$REF)[x$GT == "1|0"]
	h2[x$GT == "1|0"] <- as.character(unlist(x$ALT))[x$GT == "1|0"]
	# haplotype 1 (h1) alt allele for 0|1
	h1[x$GT == "0|1"] <- as.character(unlist(x$ALT))[x$GT == "0|1"]	
	h2[x$GT == "0|1"] <- as.character(x$REF)[x$GT == "0|1"]
	return(list(h1 = h1, h2 = h2))
}

plotHaplotypeFraction <- function(dataIn, chr = NULL, resultType = "HaplotypeRatio", colType = "Haplotypes", 
	phaseBlockCol = c("#9ad0f3", "#CC79A7"), geneAnnot = NULL, spacing = 4,  xlim = NULL,  ...) {
    if (!resultType %in% c("HaplotypeRatio", "AllelicRatio")){
      stop("plotHaplotypeFraction: resultType must be one of 'HaplotypeRatio' or 'AllelicRatio'.")
    }
    if (!colType %in% c("Haplotypes", "CopyNumber")){
      stop("plotHaplotypeFraction: plotType must be one of 'Haplotypes' or 'CopyNumber'")
    }
    lohCol.hap <- c(`0`=phaseBlockCol[1], `1`=phaseBlockCol[2])
    lohCol.titan <- c("#00FF00", "#006400", "#0000FF", "#8B0000", 
        "#006400", "#BEBEBE", "#FF0000", "#BEBEBE", 
        "#FF0000")
    names(lohCol.titan) <- c("HOMD", "DLOH", "NLOH", "GAIN", 
        "ALOH", "HET", "ASCNA", "BCNA", "UBCNA")
    dataIn <- copy(dataIn)
    colnames(dataIn)[1:2] <- c("Chr", "Position")
    #dataIn$AllelicRatio <- pmax(dataIn$refOriginal, dataIn$tumDepthOriginal - dataIn$refOriginal) / dataIn$tumDepthOriginal
    ## set alternating 0 and 1 for phaseSets
    psRuns <- rle(dataIn$PhaseSet)
    ps.id <- 1:length(psRuns$lengths) %% 2
    dataIn$phaseSet.id <- rep(ps.id, psRuns$lengths)
    #dataIn$HaplotypeCount[as.logical(dataIn$phaseSet.id)] <- dataIn$HaplotypeDepth[as.logical(dataIn$phaseSet.id)] - dataIn$HaplotypeCount[as.logical(dataIn$phaseSet.id)]
    #dataIn$HaplotypeRatio[as.logical(dataIn$phaseSet.id)] <- 1 - dataIn$HaplotypeRatio[as.logical(dataIn$phaseSet.id)]
    #dataIn$AllelicRatio <- dataIn$refOriginal / dataIn$tumDepthOriginal
    dataIn[, HaplotypeRatio.1 := HaplotypeRatio]#dataIn$HaplotypeCount / dataIn$HaplotypeDepth
    dataIn[, HaplotypeRatio.2 := 1 - HaplotypeRatio]#(dataIn$HaplotypeDepth - dataIn$HaplotypeCount) / dataIn$HaplotypeDepth
    
    if (!is.null(chr)) {
        for (i in chr) {
            dataByChr <- dataIn[Chr == i, ]
            #dataByChr <- dataByChr[dataByChr[, "TITANcall"] != "OUT", ]
            # plot the data if (outfile!=''){
            # pdf(outfile,width=10,height=6) }
            if (colType == "Haplotypes" || !"TITANcall" %in% colnames(dataByChr)){
              colors.1 <- lohCol.hap[as.character(dataByChr$phaseSet.id)]
              colors.2 <- lohCol.hap[as.character(as.numeric(!dataByChr$phaseSet.id))]
            }else{
              colors.1 <- lohCol.titan[dataByChr[, "TITANcall"]]
              colors.2 <- colors.1
            }
            
            par(mar = c(spacing, 8, 2, 2))
            # par(xpd=NA)
            if (missing(xlim)) {
                xlim <- as.numeric(c(1, dataByChr[nrow(dataByChr), Position]))
            }
            if (resultType == "HaplotypeRatio"){
              plot(dataByChr[, Position], dataByChr[, HaplotypeRatio.1], 
                  col = colors.1, 
                  pch = 16, xaxt = "n", las = 1, ylab = "Haplotype Fraction", xlim = xlim, 
                  ...)
              points(dataByChr[, Position], dataByChr[, HaplotypeRatio.2], col = colors.1, pch=16, ...)
            }else if (resultType == "AllelicRatio"){
               plot(dataByChr[, Position], dataByChr[, AllelicRatio], 
                  col = colors.1, 
                  pch = 16, xaxt = "n", las = 1, ylab = "Allelic Fraction", xlim = xlim, 
                  ...)
            }else{
              stop("Need to specify \"resultType\": HaplotypeRatio or AllelicRatio")
            }
            lines(as.numeric(c(1, dataByChr[nrow(dataByChr), Position])), rep(0.5, 2), type = "l", 
                  col = "grey", lwd = 3)
            
            if (!is.null(geneAnnot)) {
                plotGeneAnnotation(geneAnnot, i)
            }
        }
    } else {
        if (colType == "Haplotypes" || !"TITANcall" %in% colnames(dataIn)){
              colors.1 <- lohCol.hap[as.character(dataIn$phaseSet.id)]
              colors.2 <- lohCol.hap[as.character(as.numeric(!dataIn$phaseSet.id))]
            }else{
              colors.1 <- lohCol.titan[dataIn[, TITANcall]]
              colors.2 <- colors.1
        }
        
        # plot for all chromosomes
        coord <- getGenomeWidePositions(dataIn[, Chr], dataIn[, Position])
        if (resultType == "HaplotypeRatio"){
          plot(coord$posns, as.numeric(dataIn[, HaplotypeRatio.1]), 
            col = colors.1, pch = 16, 
            xaxt = "n", bty = "n", las = 1, ylab = "Haplotype Fraction", ...)
          points(coord$posns, dataIn[, HaplotypeRatio.2], col = colors.1, pch=16, ...)
        }else if (resultType == "AllelicRatio"){
          plot(coord$posns, as.numeric(dataIn[, AllelicRatio]), 
            col = colors.2, pch = 16, 
            xaxt = "n", bty = "n", las = 1, ylab = "Allelic Fraction", ...)
        }else{
              stop("Need to specify \"resultType\": HaplotypeRatio or AllelicRatio")
        }
        lines(as.numeric(c(1, coord$posns[length(coord$posns)])), 
            rep(0.5, 2), type = "l", col = "grey", 
            lwd = 3)
        plotChrLines(unique(dataIn[, Chr]), coord$chrBkpt, 
            c(-0.1, 1.1))
        
    }
}

keepChr <- function(tumour_reads, chr = c(1:22,"X","Y")){	
	tumour_reads <- tumour_reads[space(tumour_reads) %in% chr, ]
	tumour_reads <- as.data.frame(tumour_reads)
	tumour_reads$space <- droplevels(tumour_reads$space)
	tumour_reads$space <- factor(tumour_reads$space,levels=chr)
	tumour_reads <- as(tumour_reads,"RangedData")
	return(tumour_reads)
}

